port/linux/omrvmem.c

```diff
@@ -677,7 +678,14 @@ omrvmem_decommit_memory(struct OMRPortLibrary *portLibrary, void *address, uintp
 
 			if (byteAmount > 0) {
 				if (identifier->allocator == OMRPORT_VMEM_RESERVE_USED_MMAP) {
-					result  = (intptr_t)madvise((void *)address, (size_t) byteAmount, MADV_DONTNEED);
+					advice = MADV_DONTNEED;
+					/* If heap is created using shared memory with mmap, we must set advice to MADV_REMOVE, becasue
+					 * pages might not be immediately freed in a successful madvise used with MADV_DONTNEED in
+					 * shared memory case. But using MADV_REMOVE does. */
+					if (-1 != identifier->fd) {
+						advice = MADV_REMOVE;
+					}
+					result  = (intptr_t)madvise((void *)address, (size_t) byteAmount, advice);
 				} else {
 					/* need to determine what to use in the case of shmat/shmget, till then return success */
```

Comment:
I wonder if identifier->allocator should be set to OMRPORT_VMEM_RESERVE_USED_SHM when we create heap as shared (with file descriptor), and then handle decommit in this else branche

Reply:
Also fd will be set when we create heap with huge pages, in which cases we should not try this (or any decommit).

Reply:
That would also work. The only case where `OMRPORT_VMEM_RESERVE_USED_SHM` is set is in the case of huge pages where we call `shmat()` to create the heap. If we switch to use `OMRPORT_VMEM_RESERVE_USED_SHM`, we would also need to update `omrvmem_free_memory()`. In theory we would need to set both `OMRPORT_VMEM_RESERVE_USED_SHM` and `OMRPORT_VMEM_RESERVE_USED_MMAP` in case of the shared heap.

Reply:
At high level, I'm ok with this change, but @dmitripivkine knows this code better. From a discussion I had with him, my proposal is no good. 

@dmitripivkine please review.  More testing was done by @bragaigor  to prove that there are no negative effects in repetitively expand & contract heap and consequently commit & decommit (using MADV_REMOVE advice) areas of heap that have been double mapped

Reply:
We never shared mmap'ed pages before, so looks like it is necessary improvement. However I don't know enough how MADV_REMOVE works exactly. I will try to find out more about it.
@pshipton @DanHeidinga Would you please take a look as well?

Reply:
in the light of my original proposal.... should we have a hybrid allocator type _MMAP_SHM?

Reply:
+1 to Aleks' suggestion to add another allocator type to reflect shared mmap allocations. It seems using the fd to determine if shared mapping is used isn't guaranteed. There is another possible case where a /dev/zero file is used and the fd is set.